### PR TITLE
Add sessions

### PR DESCRIPTION
**Pull request by nahiluhmot**
@tlunter @djosborne 

* Add a sessions model/database table
* Add a service to create and expire sessions
* Add a controller to do the same
* Make the users controller ensure that users are signed in when modifying information

### CODE REVIEW

File: app/auth/controllers/sessions.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    # This controller handles session lifecycle.
+    class Sessions < Base
+      helpers { include Services::Sessions }
+
+      post '/sessions/?' do
+        status 201
+        hash = parse_json(req_body)
+        expires = Time.now + 30.days
+        session = create(*hash.values_at('email', 'password'), expires)
+        response.set_cookie(:session_key, session['key'])
```
Comment (tlunter) [inline app/auth/controllers/sessions.rb]: If we want it to persist between browser sessions, we should probably add a `max_age` here.

File: app/auth/controllers/sessions.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        session.to_json
+      end
+
+      delete '/sessions/?' do
+        status 204
+        key = request.cookies['session_key']
+        response.set_cookie('session_key', '')
```
Comment (tlunter) [inline app/auth/controllers/sessions.rb]: Could use `delete_cookie` instead of setting it to empty string.

File: app/auth/models/session.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+      validate :user_associated?
+
+      def as_json(opts = {})
+        super(opts.merge(except: %w(created_at updated_at)))
+      end
+
+      def expired?
+        Time.now < expires_on
+      end
```
Comment (tlunter) [inline app/auth/models/session.rb]: Usually, you'd leave this up to the cookies

**Comment by tlunter** [comment]
Looks fine for me.  Some things could probably be left up to the cookies rather than checking in the DB

File: app/auth/controllers/sessions.rb (hunk 1 above)
Comment (nahiluhmot) [inline app/auth/controllers/sessions.rb]: Will do.

File: app/auth/controllers/sessions.rb (hunk 2 above)
Comment (nahiluhmot) [inline app/auth/controllers/sessions.rb]: Good call.

File: app/auth/models/session.rb (hunk 3 above)
Comment (nahiluhmot) [inline app/auth/models/session.rb]: I'm using the `expires_on` date to tell which sessions are active and aren't.

**Comment by nahiluhmot** [comment]
@tlunter Left more stuff up to the cookies. Will squash those two commits post review.

**Comment by tlunter** [comment]
+2

**Pull request merged**